components/tools/crosshairs.js

```diff
@@ -3,306 +3,121 @@
 
     fc.tools.crosshairs = function() {
 
-        var target = null,
-            series = null,
-            xScale = d3.time.scale(),
-            yScale = d3.scale.linear(),
-            yValue = null,
-            formatH = null,
-            formatV = null,
-            active = true,
-            freezable = true,
-            padding = 2,
-            onSnap = null;
+        var event = d3.dispatch('trackingstart', 'trackingmove', 'trackingend');
 
-        var lineH = null,
-            lineV = null,
-            circle = null,
-            calloutH = null,
-            calloutV = null;
+        var crosshairs = function(selection) {
 
-        var highlight = null,
-            highlightedValue = null;
-
-        var crosshairs = function() {
-
-            var root = target.append('g')
-                .attr('class', 'crosshairs');
-
-            lineH = root.append('line')
-                .attr('class', 'crosshairs horizontal')
-                .attr('x1', xScale.range()[0])
-                .attr('x2', xScale.range()[1])
-                .attr('display', 'none');
-
-            lineV = root.append('line')
-                .attr('class', 'crosshairs vertical')
-                .attr('y1', yScale.range()[0])
-                .attr('y2', yScale.range()[1])
-                .attr('display', 'none');
-
-            circle = root.append('circle')
-                .attr('class', 'crosshairs circle')
-                .attr('r', 6)
-                .attr('display', 'none');
-
-            calloutH = root.append('text')
-                .attr('class', 'crosshairs callout horizontal')
-                .attr('x', xScale.range()[1] - padding)
-                .attr('style', 'text-anchor: end')
-                .attr('display', 'none');
-
-            calloutV = root.append('text')
-                .attr('class', 'crosshairs callout vertical')
-                .attr('y', '1em')
-                .attr('style', 'text-anchor: end')
-                .attr('display', 'none');
-        };
-
-        function mousemove() {
-
-            if (active) {
-                crosshairs.update();
-            }
-        }
-
-        function mouseout() {
-
-            if (active) {
-                crosshairs.clear();
-            }
-        }
-
-        function mouseclick() {
-
-            if (freezable) {
-                crosshairs.active(!active);
-            }
-        }
-
-        function findNearest(xTarget) {
-
-            var nearest = null,
-                dx = Number.MAX_VALUE;
-
-            series.forEach(function(data) {
-
-                var xDiff = Math.abs(xTarget - xScale(data.date));
-
-                if (xDiff < dx) {
-                    dx = xDiff;
-                    nearest = data;
+            selection.each(function() {
+                var data = this.__data__ || [];
+                if (!data.__crosshairs__) {
+                    data.__crosshairs__ = {};
+                    this.__data__ = data;
```

Comment:
I wonder if rather than adding `__crosshairs__` to `data`, it could be added directly to `this`? Some D3 components stash values in `this.__chart__`, which is a similar concept (e.g [Axis][1] and [Brush][1]).

Does this need to be in its own loop or could in be within the following `selection.each()` (Line 18)?

[1]: https://github.com/mbostock/d3/blob/4e4709ce47d5c9406290dc2a467cef08002fc9b7/src/svg/axis.js#L24
[2]: https://github.com/mbostock/d3/blob/master/src/svg/brush.js#L95